<h1>Segmenting and Clustering Neighborhoods in Toronto</h1>
<h2>Robert Leon</h2>
<h3>Coursera IBM Data Science Capstone</h3>
<p>In this exercise, I will be using the Foursquare API and scikitlearn's k-means clustering to classify Toronto neighborhoods according to the makeup of local business categories<p>

The first step will be retrieving a list of neighborhoods in Toronto. A list is available for scraping at https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [4]:
from bs4 import BeautifulSoup
import pandas as pd
import requests

In [5]:
#URL of info source
zip_codes_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

#create http request to query URL content
wiki_response = requests.get(zip_codes_url)

In [6]:
#split cell to avoid multiple http requests everytime I rerun the cell
#create BeautifulSoup object to parse response content
wikipage = BeautifulSoup(wiki_response.text)

print(wikipage.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"449ccf20-e1f6-4d16-813e-c13126a76ab2","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":969510799,"wgRevisionId":969510799,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toron

In [7]:
#use pd.read_html to create a list of dataframes, save dataframe to a variable for easy referencing 
#I used prettify() after realizing beautifulsoup returns the object as text by default, and read_html parses HTML
wiki_page_table = pd.read_html(wikipage.table.prettify())[0]

In [8]:
#Confirm success!
wiki_page_table.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [9]:
#Create mask of all rows with 'Not assigned' as borough
table_mask = wiki_page_table['Borough']!='Not assigned'
#Apply mask to wiki_page_table and save as 'clean' version to use
clean_wiki_page_table = wiki_page_table[table_mask]

In [10]:
#Pandas is not letting me view all rows, changing the 'display.max_rows' setting's value to 'None'
pd.set_option('display.max_rows', None)

#Displaying dataframe in its entirety 
clean_wiki_page_table

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
107         M9P         Etobicoke   
108         M1R       Scarborough   
109         M2R        North York   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Vic

In [22]:
#Display number of rows x number of columns for easy reference
clean_wiki_page_table.shape

(103, 3)

The next step is getting the Geo coordinates for each Zip code in thelist.

In [44]:
#import Geocoder library
import geocoder
#import json to handle geocoding api responses
import json
#import statistics package to average out boxed coordinates
import statistics

Here I create my dataframe to store each zip code's coordinates

In [129]:
#create the dataframe to store my results
lat_lon_columns=['Latitude','Longitude','Code Searched']
lat_lon_df=pd.DataFrame(columns=lat_lon_columns)

#used a for loop to loop through Postal codes, and save the center coordinates for each in a dataframe
#coordinates are presented as corners of an area, center of each area seems best if doing a radius search

for zipcode in clean_wiki_page_table['Postal Code']:
        g=geocoder.google('{}, Toronto, Ontario'.format(zipcode), key='AIzaSyDF9De9BQcP7vDsK0VxuS8ReoHOMUO-n_Q')
        latitude = statistics.mean([g.json['bbox']['northeast'][0],g.json['bbox']['southwest'][0]])
        longitude = statistics.mean([g.json['bbox']['northeast'][1],g.json['bbox']['southwest'][1]])
        lat_lon_df = lat_lon_df.append({'Latitude':latitude,'Longitude':longitude,'Code Searched':zipcode}, ignore_index=True)
        
print(lat_lon_df)

      Latitude  Longitude Code Searched
0    43.750384 -79.335351           M3A
1    43.729376 -79.312923           M4A
2    43.649954 -79.352845           M5A
3    43.723346 -79.450757           M6A
4    43.662415 -79.389786           M7A
5    43.662091 -79.528267           M9A
6    43.810176 -79.190328           M1B
7    43.748739 -79.356410           M3B
8    43.707380 -79.311904           M4B
9    43.657723 -79.378585           M5B
10   43.706510 -79.447217           M6B
11   43.650063 -79.553815           M9B
12   43.785741 -79.156456           M1C
13   43.721230 -79.350863           M3C
14   43.690582 -79.309789           M4C
15   43.651446 -79.376139           M5C
16   43.691950 -79.430395           M6C
17   43.638168 -79.578721           M9C
18   43.766170 -79.185798           M1E
19   43.678474 -79.294502           M4E
20   43.645672 -79.374024           M5E
21   43.687424 -79.450271           M6E
22   43.769966 -79.217827           M1G
23   43.706052 -79.364867           M4G


After a long time torubleshooting null resonses from the foursquare API further below, I found the above block originally logged the coordinates in the wrong order

In [131]:
#Create new DataFrame to hold merged toronto neighborhood data
t_neighborhood_columns = ['PostalCode', 'Burough', 'Neighborhood', 'Latitude', 'Longitude']
toronto_neighborhoods = pd.DataFrame(columns=t_neighborhood_columns)
lat_lon_df.set_index('Code Searched', inplace=True)

#Add the longitude and latitude coordinates to the new DF
for index, row in clean_wiki_page_table.iterrows():
    postcode = row['Postal Code']
    borough = row['Borough']
    nhood = row['Neighbourhood']
    lat = lat_lon_df.loc[postcode]['Latitude']
    lon = lat_lon_df.loc[postcode]['Longitude']
    toronto_neighborhoods = toronto_neighborhoods.append({
        'PostalCode':postcode,
        'Burough':borough,
        'Neighborhood':nhood,
        'Latitude':lat,
        'Longitude':lon
    },ignore_index=True
    )
    
#Display dataframe    
toronto_neighborhoods
    


PostalCode           Burough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.750384 -79.335351  
1                                     Victoria Village  43.729376 -79.312923  
2                     

Hiding my keys for security

In [200]:
CLIENT_ID = 'C13Q5GJLF3P3NCV2PWSK1LLHPOVJZ5QTHDAQPQTMOHYLUGBJ' # your Foursquare ID
CLIENT_SECRET = 'MHBAEXFXLIPRD35KHGXIF5PXIDMMX3HH0WVZUIBIZASGOYBN' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 50

Reusing this bit of code from a previous assignment. 
This function will build the request to the foursquare API, parse the response into JSON format, 
iterate through the nested JSON data to extract venue names, coordinates, and categories and return them as a dataframe to work with.


In [248]:
def getNearbyVenues(names, latitudes, longitudes, radius=2000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        #print(url) #- had to print url to visit directly for troubleshooting. i learned i had interpretted latitude and longitude results from geocoder in the worng order 
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostalCode','PostalCode Latitude','PostalCode Longitude','Venue','Venue Latitude','Venue Longitude','Venue Category']
    
    return(nearby_venues)

Formatting the dataframe to show every venue returned for every postal code

In [249]:
toronto_venues = getNearbyVenues(names=toronto_neighborhoods['PostalCode'],
                                   latitudes=toronto_neighborhoods['Latitude'],
                                   longitudes=toronto_neighborhoods['Longitude']
                                  )
toronto_venues.head()

M3A
M4A
M5A
M6A
M7A
M9A
M1B
M3B
M4B
M5B
M6B
M9B
M1C
M3C
M4C
M5C
M6C
M9C
M1E
M4E
M5E
M6E
M1G
M4G
M5G
M6G
M1H
M2H
M3H
M4H
M5H
M6H
M1J
M2J
M3J
M4J
M5J
M6J
M1K
M2K
M3K
M4K
M5K
M6K
M1L
M2L
M3L
M4L
M5L
M6L
M9L
M1M
M2M
M3M
M4M
M5M
M6M
M9M
M1N
M2N
M3N
M4N
M5N
M6N
M9N
M1P
M2P
M4P
M5P
M6P
M9P
M1R
M2R
M4R
M5R
M6R
M7R
M9R
M1S
M4S
M5S
M6S
M1T
M4T
M5T
M1V
M4V
M5V
M8V
M9V
M1W
M4W
M5W
M8W
M9W
M1X
M4X
M5X
M8X
M4Y
M7Y
M8Y
M8Z


PostalCode  PostalCode Latitude  PostalCode Longitude  \
0        M3A            43.750384            -79.335351   
1        M3A            43.750384            -79.335351   
2        M3A            43.750384            -79.335351   
3        M3A            43.750384            -79.335351   
4        M3A            43.750384            -79.335351   

                         Venue  Venue Latitude  Venue Longitude  \
0  Donalda Golf & Country Club       43.752816       -79.342741   
1                 Island Foods       43.745866       -79.346035   
2              Allwyn's Bakery       43.759840       -79.324719   
3         Galleria Supermarket       43.753520       -79.349518   
4              Brookbanks Park       43.751976       -79.332140   

         Venue Category  
0           Golf Course  
1  Caribbean Restaurant  
2  Caribbean Restaurant  
3           Supermarket  
4                  Park

Checking how many venues (4893), and postal codes (103) are in the data

In [203]:
toronto_venues.shape

(4893, 7)

In [204]:
toronto_venues.groupby('PostalCode').count().shape

(103, 6)

In [205]:
toronto_venues.groupby('PostalCode').count()
print('There are {} unique Venue Categories'.format(len(toronto_venues['Venue Category'].unique())))

There are 295 unique Venue Categories


Venue categories are categorical value. Since we'll be feeding these into the kMeans model, we want the values broken out into dimensions to indicate whether the dimension applies to the Venue

In [250]:
# break the Venue Category out into dummy columns 
toronto_onehot = pd.get_dummies(toronto_venues['Venue Category'], prefix="", prefix_sep="")

# add postal code column to dataframe
toronto_onehot.insert(0,'PostalCode',toronto_venues['PostalCode'])

#Having trouble seeing all columns, so changing settings to allow all to display
pd.set_option('display.max_columns', None)


I'm summarizing the above onhot table into a dataframe broken out by zip code

In [217]:
toronto_grouped = toronto_onehot.groupby('PostalCode').sum().reset_index()
toronto_grouped

PostalCode  Afghan Restaurant  African Restaurant  Airport  \
0          M1B                  0                   0        0   
1          M1C                  0                   0        0   
2          M1E                  0                   0        0   
3          M1G                  0                   0        0   
4          M1H                  0                   0        0   
5          M1J                  0                   0        0   
6          M1K                  0                   0        0   
7          M1L                  0                   0        0   
8          M1M                  0                   0        0   
9          M1N                  0                   0        0   
10         M1P                  0                   0        0   
11         M1R                  0                   1        0   
12         M1S                  0                   0        0   
13         M1T                  0                   0        0   
14         M1V                  0                   0        0   
15         M1W                  0                   0        0   
16         M1X                  0                   0        0   
17         M2H                  0                   0        0   
18         M2J                  0                   0        0   
19         M2K                  0                   0        0   
20         M2L                  0                   0        0   
21         M2M                  0                   0        0   
22         M2N                  0                   0        0   
23         M2P                  0                   0        0   
24         M2R                  0                   0        0   
25         M3A                  0                   0        0   
26         M3B                  0                   0        0   
27         M3C                  1                   0        0   
28         M3H                  0                   0        0   
29         M3J                  0                   0        0   
30         M3K                  0                   0        1   
31         M3L                  0                   0        0   
32         M3M                  0                   0        0   
33         M3N                  0                   0        0   
34         M4A                  0                   0        0   
35         M4B                  0                   0        0   
36         M4C                  0                   0        0   
37         M4E                  0                   0        0   
38         M4G                  1                   0        0   
39         M4H                  1                   0        0   
40         M4J                  0                   0        0   
41         M4K                  0                   0        0   
42         M4L                  0                   0        0   
43         M4M                  0                   0        0   
44         M4N                  0                   0        0   
45         M4P                  0                   0        0   
46         M4R                  0                   0        0   
47         M4S                  0                   0        0   
48         M4T                  0                   0        0   
49         M4V                  0                   0        0   
50         M4W                  0                   0        0   
51         M4X                  0                   0        0   
52         M4Y                  0                   0        0   
53         M5A                  0                   0        0   
54         M5B                  0                   0        0   
55         M5C                  0                   0        0   
56         M5E                  0                   0        0   
57         M5G                  0                   0        0   
58         M5H                  0                   0        0   
59         M5J                  0           

I will classify and label by zip code, rather than by neighbohood. In retrospect, a neighborhood classification may have been more practical.

In [251]:
#import the KMeans package
from sklearn.cluster import KMeans

#We will identify 5 distinct clusters
kclusters = 5

#Prep the dataframe for fitting by dropping the PostalCode column
toronto_grouped_clustering = toronto_grouped.drop(['PostalCode'],1)

#Fit the model using the new dataframe, and the predetermined number of clusters above
cluster_model = KMeans(n_clusters=kclusters, random_state=9).fit(toronto_grouped_clustering)

#print length on labels to verify it aligns with amoung of zip codes
len(cluster_model.labels_)


103

Group the zipcodes with their labels to join further down into a final table with coordinates

In [252]:
toronto_grouped['Labels'] = cluster_model.labels_
toronto_grouped[['PostalCode','Labels']]

PostalCode  Labels
0          M1B       4
1          M1C       1
2          M1E       1
3          M1G       1
4          M1H       2
5          M1J       1
6          M1K       2
7          M1L       2
8          M1M       2
9          M1N       1
10         M1P       2
11         M1R       2
12         M1S       2
13         M1T       1
14         M1V       2
15         M1W       1
16         M1X       2
17         M2H       2
18         M2J       2
19         M2K       2
20         M2L       2
21         M2M       0
22         M2N       0
23         M2P       2
24         M2R       1
25         M3A       3
26         M3B       2
27         M3C       2
28         M3H       1
29         M3J       1
30         M3K       0
31         M3L       1
32         M3M       1
33         M3N       1
34         M4A       2
35         M4B       1
36         M4C       2
37         M4E       3
38         M4G       2
39         M4H       2
40         M4J       3
41         M4K       3
42         M4L       3
43         M4M       2
44         M4N       3
45         M4P       3
46         M4R       3
47         M4S       3
48         M4T       3
49         M4V       3
50         M4W       3
51         M4X       3
52         M4Y       0
53         M5A       3
54         M5B       0
55         M5C       0
56         M5E       0
57         M5G       0
58         M5H       0
59         M5J       0
60         M5K       0
61         M5L       0
62         M5M       3
63         M5N       3
64         M5P       3
65         M5R       3
66         M5S       3
67         M5T       3
68         M5V       3
69         M5W       0
70         M5X       0
71         M6A       0
72         M6B       2
73         M6C       3
74         M6E       3
75         M6G       3
76         M6H       3
77         M6J       3
78         M6K       3
79         M6L       1
80         M6M       1
81         M6N       3
82         M6P       3
83         M6R       3
84         M6S       3
85         M7A       0
86         M7R       0
87         M7Y       3
88         M8V       3
89         M8W       2
90         M8X       3
91         M8Y       3
92         M8Z       2
93         M9A       1
94         M9B       1
95         M9C       1
96         M9L       1
97         M9M       1
98         M9N       1
99         M9P       1
100        M9R       1
101        M9V       1
102        M9W       1

I thought i would improve on the class example by feeding the whole count of nearby venue categories rather than just the top 10. I found the data a bit overwhelming to process. I settled on analyzing their average and sorting by columns rather than rows. In the end, I checked only the top 10 categories anyway. I had gone a long about way to do it the same way as the assignment. Listen to your teachers!

In [240]:
df_to_sort = toronto_grouped.groupby('Labels').mean()


for row, col in df_to_sort.iterrows():
    print(col.sort_values(ascending=False)[:10])
    

Coffee Shop            3.000000
Hotel                  2.214286
Café                   2.000000
Japanese Restaurant    1.785714
Clothing Store         1.785714
Restaurant             1.571429
Park                   1.571429
Plaza                  1.285714
Gym                    1.142857
Italian Restaurant     1.071429
Name: 0, dtype: float64
Coffee Shop              5.208333
Pizza Place              2.750000
Bank                     2.708333
Sandwich Place           2.458333
Fast Food Restaurant     2.250000
Pharmacy                 2.125000
Gas Station              1.708333
Grocery Store            1.541667
Vietnamese Restaurant    1.125000
Park                     1.000000
Name: 1, dtype: float64
Coffee Shop                  3.379310
Grocery Store                1.586207
Bakery                       1.310345
Chinese Restaurant           1.206897
Bank                         1.172414
Burger Joint                 1.068966
Restaurant                   1.068966
Japanese Restaurant       

After much thought, I'm going to hypothesize on the nature of the clusters<br/>
Cluster 0 - Tourist/commercial center because Hotels and plazas<br/>
Cluster 1 - High Density Residential - banks, fast food, more pharmacy than grocery stores<br/>
Cluster 2 - Less Dense residential - not many venues,characterized by grocery stores, banks, restuarants<br/>
Cluster 3 - Nightlife are - date attractions like restuarants, cafes, ice cream, sushi, bars<br/>
Cluster 4 - Far out zoo area - fast food, zoo, gas stations<br/>

In [243]:
toronto_merged = toronto_neighborhoods.join(toronto_grouped[['PostalCode','Labels']].set_index('PostalCode'),on='PostalCode')

In [242]:
toronto_merged

PostalCode           Burough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
0                                            Parkwoods  43.750384 -79.335351   
1                                     Victoria Village  43.729376 -79.312923   
2                  

Iterate through the data and plot the rows on a map

In [247]:
import folium
import numpy as np

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

#Toronto Coordinates to initialize map
latitude = 43.6532
longitude = -79.3832

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['PostalCode'], toronto_merged['Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters